framework/Web/Services/TPageService.php

```diff
@@ -480,7 +482,23 @@ protected function createPage($pagePath)
 		$hasClassFile = is_file($path . Prado::CLASS_FILE_EXT);
 
 		if (!$hasTemplateFile && !$hasClassFile) {
-			throw new THttpException(404, 'pageservice_page_unknown', $pagePath);
+			$paths = $this->onAdditionalPagePaths($pagePath);
```

Comment:
Why use an event here? this means that for every single page each plugin will have to redeclare the additional path where to load its pages from.
Maybe it would be more convenient for TApplication/TPageService to have an array $additionalPageBasePaths and a couple of methods  getAdditionalPageBasePaths() addAdditionalPageBasePaths() so that a plugin can register its additional path(s) once when being loaded.
What do you think?

Reply:
Why use an event here?

Events return an array of results. (dynamic events return a single value, I tried this as a dynamic event, a regular event is far better).

Each page would NOT have to redeclare an additional path. here is why.  The one handler for one plugin, but all its pages, would look like this:

> 	public function additionalPagePaths($service, $pagePath)
> 	{
> 		 return $this->getPluginPagesPath() . DIRECTORY_SEPARATOR . strtr($pagePath, '.', DIRECTORY_SEPARATOR);
> 	}

This resolves the major issue you raise.  It would give the plugin page path for "Admin.MyPluginPages" by creating it on the fly.

Adding a handler is about the same as registering an additionalPageBasePath.  this removes additional variables for the class.  reducing complexity.  It localizes the changes and localizes additional functionality. And gives power to the plugin to handle its own pages the way the developer wants.

Reply:
this is documented in the php doc for the event

Reply:
If an array of BasePaths is better, in your view, then I'll make the change as per your suggestion.

Reply:
Wordpress, yii, Drupal and other platforms use events in the kind of ways that onAdditionalPagePaths is used here.  It's one use cases for the change to raiseEvent to return handler data in an array.

Reply:
You have a point here. Sorry I ended this review at about midnight, somehow I convinced myself that createPage() could be called more then once is a single page request.. I woke up this morning asking myself why :)

Reply:
I've had similar moments.  we all have.  :)